sidecar/src/agentic/tool/mcp/integration_tool.rs

```diff
@@ -0,0 +1,195 @@
+                    .map(|s| s.to_string());
+                let schema = t.get("schema").cloned();
+                tool_descriptors.push(ToolDescriptor {
+                    name,
+                    description,
+                    schema,
+                });
+            }
+
+            server_list.push(ServerTools {
+                server_name: server_name.clone(),
+                tools: tool_descriptors,
+            });
+        }
+        Ok(MCPIntegrationToolResponse::ToolList(ToolListResponse {
+            servers: server_list,
+        }))
+    }
+
+    async fn call_tool(
+        &self,
+        server_name: &str,
+        tool_name: &str,
+        arguments: Value,
+    ) -> Result<MCPIntegrationToolResponse, ToolError> {
+        let client = self.servers.get(server_name).ok_or_else(|| {
+            ToolError::InvalidInput(format!("Server '{}' not found", server_name))
+        })?;
+        let res = client
+            .call_tool(tool_name, arguments)
+            .await
+            .map_err(|e| ToolError::InvocationError(format!("call_tool failed: {}", e)))?;
+
+        Ok(MCPIntegrationToolResponse::ToolCall(ToolCallResponse {
+            result: res,
+        }))
+    }
+
+    async fn handle_query(
+        &self,
+        query: MCPIntegrationToolQuery,
+    ) -> Result<MCPIntegrationToolResponse, ToolError> {
+        match query.action.as_str() {
+            "list" => self.list_all_tools().await,
+            "call" => {
+                let server_name = query.server_name.as_ref().ok_or_else(|| {
+                    ToolError::InvalidInput("Missing 'server_name' for call action".to_string())
+                })?;
+                let tool_name = query.tool_name.as_ref().ok_or_else(|| {
+                    ToolError::InvalidInput("Missing 'tool_name' for call action".to_string())
+                })?;
+                self.call_tool(server_name, tool_name, query.arguments.clone())
+                    .await
+            }
+            _ => Err(ToolError::InvalidInput("Unknown action".to_string())),
+        }
+    }
+}
+
+#[async_trait]
+impl Tool for MCPIntegrationToolBroker {
+    async fn invoke(&self, input: ToolInput) -> Result<ToolOutput, ToolError> {
+        let query = match input {
+            ToolInput::MCPIntegrationTool(q) => q,
+            _ => {
+                return Err(ToolError::InvalidInput(
+                    "Expected MCPIntegrationTool input".to_string(),
+                ))
+            }
+        };
+
+        let response = self.handle_query(query).await?;
+        Ok(ToolOutput::MCPIntegration(response))
+    }
+
+    fn tool_description(&self) -> String {
+        "The MCP Integration tool: Use 'action':'list' to list all servers & tools, 'action':'call' to invoke a tool.".to_string()
+    }
+
+    fn tool_input_format(&self) -> String {
+        r#"{"action":"list"} or {"action":"call","server_name":"string","tool_name":"string","arguments":{}}"#.to_string()
+    }
```

Comment:
this format is generally used for parsing xml and not json which might lead to errors over here